src/auxinfo/proof.rs

```diff
@@ -143,3 +143,193 @@ impl AuxInfoProof {
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        paillier::prime_gen,
+        protocol::SharedContext,
+        utils::{k256_order, testing::init_testing, CurvePoint},
+        ParticipantIdentifier,
+    };
+    use rand::random;
+
+    use super::*;
+
+    pub(crate) fn generate_shared_context() -> SharedContext {
+        let mut rng = init_testing();
+        let sid = Identifier::random(&mut rng);
+        let participant = ParticipantIdentifier::random(&mut rng);
+        let participant2 = ParticipantIdentifier::random(&mut rng);
+        let participants = vec![participant, participant2];
+        let generator = CurvePoint::GENERATOR;
+        let order = k256_order();
+        SharedContext::gen_shared_context(sid, participants, generator, order)
+    }
+
+    #[test]
+    fn auxinfo_proof_verifies() -> Result<()> {
+        let mut rng = init_testing();
+        let sid = Identifier::random(&mut rng);
+        let rho = random();
+        let setup_params = VerifiedRingPedersen::gen(&mut rng, &())?;
+        let (p, q) = prime_gen::get_prime_pair_from_pool_insecure(&mut rng).unwrap();
+        let modulus = &p * &q;
+        let shared_context = &generate_shared_context();
+        let proof = AuxInfoProof::prove(
+            &mut rng,
+            shared_context,
+            sid,
+            rho,
+            &setup_params,
+            &modulus,
+            &p,
+            &q,
+        )?;
+        assert!(proof
+            .verify(shared_context, sid, rho, &setup_params, &modulus)
+            .is_ok());
+        Ok(())
+    }
+
+    #[test]
+    fn each_constituent_proof_must_be_valid() -> Result<()> {
```

Comment:
I can definitely write a random_auxinfo_proof function generically. But for the purposes of this test particularly, we need to construct the proof from the 2 constituent parts and then mix and match between the pimod and pifac from 2 different separately working proofs to demonstrate that it does not work. So such a function would not help here unless I write 2 separate random_pimod_proof and random_pifac_proof. Shall I do such a thing?

Reply:
If you make two random auxinfo proofs, you should be able to directly access their fields in the tests to construct the invalid one. This is because the tests are in a submodule of the auxinfo proof, so they can access private methods that you wouldn't be able to touch outside of this module.

```rust
let proof1: AuxInfoProof = random_auxinfo_proof();
let proof2: AuxinfoProof = random_auxinfo_proof();

let mix_one = AuxInfoProof {
    pifac: proof1.pifac,
    pimod: proof2.pimod,
};

let mix_two = AuxInfoProof {
    pifac: proof2.pifac,
    pimod = proof1.pimod,
};
```